src/main/java/org/spongepowered/common/data/processor/data/entity/GameModeDataProcessor.java

```diff
@@ -0,0 +1,125 @@
+import org.spongepowered.api.data.DataHolder;
+import org.spongepowered.api.data.DataTransactionBuilder;
+import org.spongepowered.api.data.DataTransactionResult;
+import org.spongepowered.api.data.key.Key;
+import org.spongepowered.api.data.key.Keys;
+import org.spongepowered.api.data.manipulator.immutable.entity.ImmutableGameModeData;
+import org.spongepowered.api.data.manipulator.mutable.entity.GameModeData;
+import org.spongepowered.api.data.merge.MergeFunction;
+import org.spongepowered.api.data.value.BaseValue;
+import org.spongepowered.api.data.value.immutable.ImmutableValue;
+import org.spongepowered.api.entity.player.Player;
+import org.spongepowered.api.entity.player.gamemode.GameMode;
+import org.spongepowered.common.Sponge;
+import org.spongepowered.common.data.ImmutableDataCachingUtil;
+import org.spongepowered.common.data.manipulator.immutable.entity.ImmutableSpongeGameModeData;
+import org.spongepowered.common.data.manipulator.mutable.entity.SpongeGameModeData;
+import org.spongepowered.common.data.processor.common.AbstractSpongeDataProcessor;
+import org.spongepowered.common.data.util.DataUtil;
+import org.spongepowered.common.data.value.immutable.ImmutableSpongeValue;
+
+public class GameModeDataProcessor extends AbstractSpongeDataProcessor<GameModeData, ImmutableGameModeData> {
+
+    @Override
+    public boolean supports(DataHolder dataHolder) {
+        return dataHolder instanceof Player;
+    }
+
+    @Override
+    public Optional<GameModeData> from(DataHolder dataHolder) {
+        if (supports(dataHolder)) {
+            if (dataHolder instanceof EntityPlayerMP) {
+                Optional.of(new SpongeGameModeData(
+                        (GameMode) ((Object) ((EntityPlayerMP) dataHolder).theItemInWorldManager.getGameType())));
+            }
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<GameModeData> fill(DataHolder dataHolder, GameModeData manipulator, MergeFunction overlap) {
+        if (dataHolder instanceof EntityPlayerMP) {
+            final GameModeData original = from(dataHolder).get();
+            return Optional.of(manipulator.set(Keys.GAME_MODE, overlap.merge(manipulator, original).type().get()));
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public Optional<GameModeData> fill(DataContainer container, GameModeData gameModeData) {
+        final String modeId = DataUtil.getData(container, Keys.GAME_MODE, String.class);
+        final Optional<GameMode> optional = Sponge.getSpongeRegistry().getType(GameMode.class, modeId);
+        if (optional.isPresent()) {
+            gameModeData.set(Keys.GAME_MODE, optional.get());
+        }
+        return Optional.absent();
+    }
+
+    @Override
+    public DataTransactionResult set(DataHolder dataHolder, GameModeData manipulator, MergeFunction function) {
+        if (dataHolder instanceof EntityPlayerMP) {
+            final GameMode oldMode =
```

Comment:
This could be simplified by getting the gamemode from the `GameModeData oldData`.